### PR TITLE
Native Auth SignIn, Signup and SSPR features

### PR DESCRIPTION
**Pull request by SaurabhMSFT**
This PR is to merge the SignIn, SIgnup and SSPR code into dev branch. This code was previously reviewed in #1897,  #1919, #1949 and #1950 .

A technical overview of Native Auth is available at https://dev.azure.com/IdentityDivision/DevEx/_git/AuthLibrariesApiReview?path=/%5BAndroid%5D%20Native%20authentication/technical_overview.md&version=GBsammy/native-authentication&_a=preview

A companion PR in common repo is  https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2233

### CODE REVIEW

File: changelog [hunk 1]
```diff
@@ -1,6 +1,10 @@
 MSAL Wiki : https://github.com/AzureAD/microsoft-authentication-library-for-android/wiki
 vNext
 ----------
+
+Version 4.10.0
+----------
+- [PATCH] Update common @16.2.0
```
Comment (fadidurah) [inline changelog]: Interesting that this is showing up here

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java [hunk 2]
```diff
@@ -959,6 +1119,43 @@ public void onCancel() {
...
+        }
+
+        if (challengeTypes != null) {
+            config.setChallengeTypes(challengeTypes);
+        }
+
+        // Check whether account mode is set to SINGLE
+        validateAccountModeConfiguration(config);
+
+        // TODO check with MSAL team
```
Comment (fadidurah) [inline msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:1150]: This seems like an old TODO, what is it referring to?

File: msal/src/main/java/com/microsoft/identity/client/AccountAdapter.java [hunk 3]
```diff
@@ -82,7 +82,10 @@ public List<ICacheRecord> filter(@NonNull final List<ICacheRecord> records) {
             for (final ICacheRecord cacheRecord : records) {
                 final String acctHomeAccountId = cacheRecord.getAccount().getHomeAccountId();
                 final String acctLocalAccountId = cacheRecord.getAccount().getLocalAccountId();
-                if (acctHomeAccountId.contains(acctLocalAccountId)) {
+
+                // TODO: check the logic on broker side to make sure this value is NOT null.
```
Comment (fadidurah) [inline msal/src/main/java/com/microsoft/identity/client/AccountAdapter.java]: Is this TODO relevant?

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java [hunk 4]
```diff
@@ -1079,15 +1277,15 @@ private void initializeApplication() throws MsalClientException {
 
         // Since network request is sent from the sdk, if calling app doesn't declare the internet
         // permission in the manifest, we cannot make the network call.
-        checkInternetPermission();
+        checkInternetPermission(mPublicClientConfiguration);
```
Comment (fadidurah) [inline msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:1280]: Did native auth make this change?

**Review by fadidurah — APPROVED** [review]
LGTM, looks like you need a dev update though

File: changelog (hunk 1 above)
Comment (SaurabhMSFT) [reply changelog]: merging with dev resolved it.

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java (hunk 4 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:1280]: yes, it was done after a class method was converted to static method. it was merged in https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1897 PR.

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java (hunk 2 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:1150]: The comment was on if we should set the shared device flag to false. We had discussed in one of our previous chats and the result was that we should remove it. I have removed the TODO too.

File: msal/src/main/java/com/microsoft/identity/client/AccountAdapter.java (hunk 3 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/client/AccountAdapter.java]: This change has gone away after merging with dev branch

**Review by shahzaibj — APPROVED** [review]

**Pull request merged**